### PR TITLE
feat(ci): migrate SBOM attach to signed cosign attestations (#161)

### PR DESCRIPTION
**Pull request by Guy-Ghis**
## 1. Summary

This PR changes:

- `.github/workflows/publish-charts-oci.yml` — the `Sign charts (cosign) + attest SBOM` step now:
  - runs an **unconditional `cosign sign`** on every chart (uniform `.sig` baseline, `cosign verify` works for all charts), and
  - additionally runs **`cosign attest --type cyclonedx`** (a signed in-toto attestation carrying the CycloneDX SBOM as its predicate) for charts whose SBOM renders (adds a `.att`, verified with `cosign verify-attestation`).
- `docs/patterns/supply-chain.md` — consumer verification documents **both** surfaces (`cosign verify` for all, `cosign verify-attestation --type cyclonedx` for SBOM-bearing charts), uses bare `ghcr.io` refs (no `oci://` for cosign), and documents the **tag-vs-digest** rationale.
- `docs/adr/0132-sbom-and-cosign-sign-oci-charts.md` + `docs/adr/README.md` — reflect the attestation approach.

It solves:

- **Source of truth:** ticket [#161](https://github.com/ADORSYS-GIS/ai-helm/issues/161) (SBOM + keyless cosign signing for OCI charts), under story #136 / epic #144. The previous implementation emitted cosign's deprecation warning (`SBOM attachments are deprecated ... use SBOM attestations`) and attached the SBOM **unsigned** (`cosign attach sbom` doesn't sign). Migrating to `cosign attest` makes the SBOM **tamper-evident**, removes the deprecation warning, and keeps a `.sig` on every chart so the verify surface stays uniform.

---

## 2. Intent

The intent of this PR is:

> `cosign attach sbom` is deprecated (removal announced for after 2024-02-22) and, critically, does **not** sign the SBOM — it is a separate, unsigned accessor from the chart signature. `cosign attest --predicate <sbom> --type cyclonedx` produces a **signed in-toto attestation** whose predicate is the CycloneDX SBOM, binding SBOM to signature in one tamper-evident step. The workflow still runs a plain `cosign sign` on **every** chart first, so the `.sig` provenance baseline (and `cosign verify`) remains uniform regardless of SBOM renderability — a transient attest failure can never leave a chart unsigned. This preserves the ADR-0132 provenance guarantee (keyless, deliberate hard-fail) while eliminating the deprecation warning and making the SBOM verifiable. Reference: [#161](https://github.com/ADORSYS-GIS/ai-helm/issues/161).

---

### In Scope

- Migrate the per-chart cosign step to `cosign attest --type cyclonedx` (signed in-toto SBOM attestation), added on top of the unconditional `cosign sign`.
- Keep a `.sig` on every chart so the provenance/verify surface is uniform across all charts.
- Update consumer-facing docs: `cosign verify` (all) + `cosign verify-attestation --type cyclonedx` (SBOM charts), bare `ghcr.io` refs, and the tag-vs-digest decision.
- Keep `tools/oci-chart-sbom.sh` unchanged (still emits the CycloneDX SBOM; only how it's attached/signed changed).

### Out of Scope

- The benign `cosign login` "credentials stored unencrypted" warning (ephemeral runner, scoped short-lived token) — no change.
- In-cluster admission-time enforcement — already documented out of scope in ADR-0132.

---

## 4. Verification

I verified this change by:

- [x] Running automated tests
- [x] Running manual tests
- [ ] Checking logs
- [ ] Checking metrics
- [ ] Testing error cases
- [ ] Testing permissions/security behavior
- [ ] Testing rollback or failure behavior, if relevant

Commands run:

```bash
# workflow YAML is well-formed (mikefarah yq, as used in CI)
yq e '.' .github/workflows/publish-charts-oci.yml >/dev/null

# shell syntax of the migrated sign step (extracted, verified)
bash -n /tmp/opencode/signstep2.sh

# SBOM generator still emits valid CycloneDX 1.5 (unchanged)
tools/oci-chart-sbom.sh charts/mongodb-backup t-mongo /tmp/sbom.json
jq -c '{bomFormat,specVersion,comps:(.components|length)}' /tmp/sbom.json
```

Results:

```text
YAML OK
shell syntax OK
{"bomFormat":"CycloneDX","specVersion":"1.5","comps":2}
```

> The full `cosign attest` run executes only in the GitHub Actions publish workflow (needs GHCR creds + OIDC). Final run evidence (`.sig` + `.att` created, `cosign verify` + `cosign verify-attestation` passing) to be recorded on #161 once this merges and a publish run completes.

---

## 5. Screenshots / Evidence

* Previous behaviour (from the #161 run): `WARNING: SBOM attachments are deprecated ... use SBOM attestations` and `Uploading SBOM file ... mediaType [text/spdx+json]` (unsigned, mislabelled media type).
* This change: every chart gets a `.sig` (`cosign verify`), and charts with a renderable SBOM additionally get a **signed** `.att` attestation with a `cyclonedx`-typed predicate (`cosign verify-attestation`) — no deprecation warning, SBOM tamper-evident.
* Pending (post-merge): a `publish-charts-oci` run + `cosign verify` / `cosign verify-attestation` output recorded on #161.

---

## 6. Risk Assessment

Risk level:

* [x] Low
* [ ] Medium
* [ ] High

Potential risks:

* `cosign attest` writes a `.att` accessor in addition to `.sig`; consumers must use `cosign verify-attestation` (not `cosign download sbom`) for the SBOM predicate — docs updated in this PR.
* A failed attestation hard-fails the step; however the chart is already `.sig`-signed beforehand, so it is never left unsigned, and charts are pushed before this step so a failure can't strand a release.

Mitigation:

* Consumer docs updated to the attestation syntax (and both verify surfaces documented).
* Signing runs before attestation, and re-runs are idempotent.

---

## 7. AI Usage Declaration

AI was used for:

- [x] Understanding existing code
- [x] Generating code
- [x] Refactoring
- [ ] Generating tests
- [x] Drafting documentation
- [x] Reviewing the diff
- [ ] Not used

Human verification:

- [x] I understand every meaningful change in this PR
- [x] I checked generated code manually
- [x] I checked generated tests manually
- [x] I removed unsupported AI assumptions
- [x] I accept responsibility for this PR

---

## 8. Reviewer Focus

Please focus your review on:

- [x] Correctness
- [x] Architecture
- [x] Security
- [ ] Performance
- [ ] Tests
- [ ] Maintainability
- [x] Product intent
- [x] Edge cases

### CODE REVIEW

**Comment by Guy-Ghis** [comment]
@lightbridge-assistant review critically and detaily for prod readiness

File: .github/workflows/publish-charts-oci.yml [hunk 1]
```diff
@@ -211,25 +212,25 @@ jobs:
...
                && [[ -s "$sbom" ]] && jq -e . "$sbom" >/dev/null 2>&1; then
-              echo "  attaching SBOM: $(jq -c '{components:(.components|length)}' "$sbom")"
-              cosign attach sbom --sbom "$sbom" "$COSIGN_REPO/$c:$ver" || \
-                echo "  (warn) sbom attach failed for $c"
+              echo "  attesting SBOM: $(jq -c '{components:(.components|length)}' "$sbom")"
+              # DELIBERATE hard-fail: a failed attestation means the chart was not
+              # signed at all, silently regressing the ADR-0132 provenance guarantee.
+              # Runs after the chart is already pushed, so it cannot strand a release —
+              # it only marks the run red for an idempotent re-run.
+              cosign attest --yes --predicate "$sbom" --type cyclonedx "$COSIGN_REPO/$c:$ver"
```
[the remark this answers is not in the document]
Comment (Guy-Ghis) [reply .github/workflows/publish-charts-oci.yml:230]: Addressed. The step now runs an **unconditional `cosign sign` on every chart first** (uniform `.sig` baseline, so `cosign verify` works for all charts), then `cosign attest --type cyclonedx` as an **additional** `.att` only when an SBOM renders. So:
- every chart carries a `.sig` (previously the SBOM branch dropped it) — uniform `cosign verify` surface restored;
- SBOM-bearing charts additionally carry a signed `.att` — `cosign verify-attestation` works for those;
- signing runs *before* the attest, so a transient attest failure never leaves a chart unsigned.

`supply-chain.md` now documents both verify surfaces (`cosign verify` for all, `cosign verify-attestation` for SBOM charts). Pushed to `feat/161-cosign-sbom-attestation` (`bc1359ba`).

[the remark this answers is not in the document]
Comment (Guy-Ghis) [reply .github/workflows/publish-charts-oci.yml:230]: Addressed — the two branches no longer diverge. Every chart now gets an unconditional `cosign sign` (uniform `.sig`), so `cosign verify` succeeds on **all** charts, and SBOM-bearing charts additionally get a signed `.att` (`cosign verify-attestation --type cyclonedx`). The former unconditional `cosign sign` is restored (before the attest), so a transient attest failure leaves the chart `.sig`-signed, never unsigned. `supply-chain.md` documents both commands and which chart class each covers. Pushed as `bc1359ba`.
